src/shims/x86/ssse3.rs

```diff
@@ -0,0 +1,206 @@
+
+                let (left, left_len) = this.operand_to_simd(left)?;
+                let (right, right_len) = this.operand_to_simd(right)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(dest_len, left_len);
+                assert_eq!(dest_len, right_len);
+
+                for i in 0..dest_len {
+                    let left = this.read_scalar(&this.project_index(&left, i)?)?.to_i16()?;
+                    let right = this.read_scalar(&this.project_index(&right, i)?)?.to_i16()?;
+                    let dest = this.project_index(&dest, i)?;
+
+                    let res = (i32::from(left).checked_mul(right.into()).unwrap() >> 14)
+                        .checked_add(1)
+                        .unwrap()
+                        >> 1;
+
+                    this.write_scalar(Scalar::from_i16(res.try_into().unwrap()), &dest)?;
+                }
+            }
+            // Used to implement the _mm_sign_epi{8,16,32} functions.
+            // Negates elements from `left` when the its corresponding from
+            // `right` is negative. If an element from `right` is zero, zero
+            // is writen to the corresponding output element.
+            "psign.b.128" | "psign.w.128" | "psign.d.128" => {
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (left, left_len) = this.operand_to_simd(left)?;
+                let (right, right_len) = this.operand_to_simd(right)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(dest_len, left_len);
+                assert_eq!(dest_len, right_len);
+
+                for i in 0..dest_len {
+                    let dest = this.project_index(&dest, i)?;
+                    let left = this
+                        .read_scalar(&this.project_index(&left, i)?)?
+                        .to_int(dest.layout.size)?;
+                    let right = this
+                        .read_scalar(&this.project_index(&right, i)?)?
+                        .to_int(dest.layout.size)?;
+
+                    let res = if right == 0 {
+                        0
+                    } else if right < 0 {
+                        if left == dest.layout.size.signed_int_min() {
+                            // Make sure `Scalar::from_int` does not get an overflowed value.
+                            //
+                            // For example, with 16-bit value -0x8000, negating after sign
+                            // extension to i128 will produce 0x8000. However, `Scalar::from_int`
+                            // will fail because 0x8000 is outside the range of signed 16-bit
+                            // integers. If the negation had been done directly with wrapping 16-bit
+                            // arithmetic, the result would be -0x8000, so we simulate that.
+                            left
```

Comment:
Can you avoid that by using `wrapping_binary_op` instead, multiplying `left` with `Scalar::from_int(right.to_int().signum())`?